nistats/glm.py

```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
+    """ compute the specified contrast given an estimated glm
+
+    Parameters
+    ----------
+    labels : array of shape (n_voxels),
+        a map of values on voxels used to identify the correseponding the model
+
+    results : dictionary,
+        keys correspond to the differnt labels
+        values are RegressionResults instances corresponding to the voxels
+
+    con_val : numpy.ndarray of shape (p) or (q, p)
+        where q = number of contrast vectors and p = number of regressors
+
+    contrast_type : {None, 't', 'F' or 'tmin-conjunction'}, optional
+        type of the contrast.  If None, then defaults to 't' for 1D
+        `con_val` and 'F' for 2D `con_val`
+
+    Returns
+    -------
+    con : Contrast instance,
+        Yields the statistics of the contrast (effects, variance, p-values)
+    """
+    con_val = np.asarray(con_val)
+    dim = 1
+    if con_val.ndim > 1:
+        dim = con_val.shape[0]
+
+    if contrast_type is None:
+        if dim == 1:
+            contrast_type = 't'
         else:
-            self.labels_ = np.zeros(Y.shape[1])
-            self.results_ = {0.0: ols_result}
+            contrast_type = 'F'
+
+    if contrast_type not in ['t', 'F']:
+        raise ValueError('Unknown contrast type: %s' % contrast_type)
+
+    effect_ = np.zeros((dim, labels.size), dtype=np.float)
+    var_ = np.zeros((dim, dim, labels.size), dtype=np.float)
+    if contrast_type == 't':
+        for label_ in regression_result.keys():
+            resl = regression_result[label_].Tcontrast(con_val)
+            effect_[:, labels == label_] = resl.effect.T
+            var_[:, :, labels == label_] = (resl.sd ** 2).T
+    else:
+        for label_ in regression_result.keys():
+            resl = regression_result[label_].Fcontrast(con_val)
+            effect_[:, labels == label_] = resl.effect
+            var_[:, :, labels == label_] = resl.covariance
+
+    dof_ = regression_result[label_].df_resid
+    return Contrast(effect=effect_, variance=var_, dof=dof_,
+                    contrast_type=contrast_type)
+
+
+class FirstLevelGLM(BaseEstimator, TransformerMixin, CacheMixin):
+    """ Implementation of the General Linear Model for Single-session fMRI data
+
+    mask: Niimg-like object, instance of NiftiMasker or MultiNiftiMasker,
+     optional
+        Mask to be used on data. If an instance of masker is passed,
+        then its mask will be used. If no mask is given,
+        it will be computed automatically by a MultiNiftiMasker with default
+        parameters.
+
+    target_affine: 3x3 or 4x4 matrix, optional
+        This parameter is passed to image.resample_img. Please see the
+        related documentation for details.
+
+    target_shape: 3-tuple of integers, optional
+        This parameter is passed to image.resample_img. Please see the
+        related documentation for details.
+
+    low_pass: False or float, optional
+        This parameter is passed to signal.clean. Please see the related
```

Comment:
Maybe nilearn.signal.clean? A few more occurences in the neighbouring lines too.